[DependencyInjection] Update ContainerBag to throw ParameterNotFoundException instead of InvalidArgumentException

WedgeSama opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 6.2
| Bug fix?      | no
| New feature?  | no
| Deprecations? | no
| Tickets       | 
| License       | MIT
| Doc PR        | 

The `Symfony\Component\DependencyInjection\ParameterBag\ContainerBag` class implement `Symfony\Component\DependencyInjection\ParameterBag\ParameterBagInterface` and the phpdoc of `get` method said to throw `Symfony\Component\DependencyInjection\ParameterBag\ParameterNotFoundException` if the parameter is not defined.

But, this `get` method call the `getParameter` method from its property `container` of class `Symfony\Component\DependencyInjection\Container`. This one throw `Symfony\Component\DependencyInjection\Exception\InvalidArgumentException` if the parameter is not defined.

It is more appropriate to throw `ParameterNotFoundException` instead of basic `InvalidArgumentException`.
There is no break backward because `ParameterNotFoundException` extends from `InvalidArgumentException`.

I let the existing test which check that the exception throw is `InvalidArgumentException` (check break backward compatibility) and add a new test only to check the usage of `ParameterNotFoundException`.

ogizanagi commented on src/Symfony/Component/DependencyInjection/ParameterBag/ContainerBag.php:
```diff
@@ -40,7 +42,11 @@ public function all()
     {
-        return $this->container->getParameter($name);
+        try {
+            return $this->container->getParameter($name);
```
I think `Container::getParameter` should be the one for which the docblock must be fixed, to indicates it throws a `ParameterNotFoundException` rather than a simple `InvalidArgumentException`.

WedgeSama replied:
Yes, i agree too. I'll update this.
Maybe directly update `ContainerInterface` too?

WedgeSama replied:
It maybe more complex than its seems (ive missed something at first). I need your opinion.

The basic `Container` throw the right exception `ParameterNotFoundException` as excepted, it is just the phpdoc that is not right.

But, there really is one container class that throw `InvalidArgumentException` instead of `ParameterNotFoundException`, it is the one generated by the `PhpDumper`, look at [PhpDumper#1531](https://github.com/symfony/symfony/blob/5.4/src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php#L1531=), [PhpDumper#1580](https://github.com/symfony/symfony/blob/5.4/src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php#L1580=) and [PhpDumper#1539](https://github.com/symfony/symfony/blob/5.4/src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php#L1539=).

I have update the PR to really fixthe throw this time. I have updated the fixtures files for dumper and tests passed.

Is this ok?

ogizanagi replied:
These changes look right to me. We're loosing the "dynamic parameter" info from the original exception message, but I'm not convinced it matters much.

WedgeSama replied:
To prevent loosing the info of dynamic parameter, i can create a new exception that extend `ParameterNotFoundException`, something like `MissingDynamicParameterException`.

Commit:
[DependencyInjection] Update PhpDumper generated Container to throw ParameterNotFoundException instead of InvalidArgumentException (more appropriate and follow throw in phpdoc)

The method `getParameter` in the container generated by the `PhpDumper` throw `InvalidArgumentException` instead of `ParameterNotFoundException` if the parameter is not defined.

This commit fix it + update tests.

Also update the basic `Container` and `ContainerInterface` throw phpdoc with the right exception `ParameterNotFoundException`.

nicolas-grekas commented:
This should target 6.2 as it's not a bug fix but an improvement.

WedgeSama commented:
Ok, i will target 6.2

fabpot reviewed (changes requested):
LGTM.
Can you rebase on current 6.2?

fabpot reviewed (approved):
I've done the rebase and resolved the conflicts.

fabpot commented:
Thank you @WedgeSama.
